libraries/stdlib/jvm/src/kotlin/text/StringNumberConversionsJVM.kt

```diff
@@ -239,35 +239,215 @@ public fun String.toBigDecimalOrNull(): java.math.BigDecimal? =
 
-/**
- * Recommended floating point number validation RegEx from the javadoc of `java.lang.Double.valueOf(String)`
- */
-private object ScreenFloatValueRegEx {
-    @JvmField val value = run {
-        val Digits = "(\\p{Digit}+)"
-        val HexDigits = "(\\p{XDigit}+)"
-        val Exp = "[eE][+-]?$Digits"
-
-        val HexString = "(0[xX]$HexDigits(\\.)?)|" + // 0[xX] HexDigits ._opt BinaryExponent FloatTypeSuffix_opt
-                "(0[xX]$HexDigits?(\\.)$HexDigits)"  // 0[xX] HexDigits_opt . HexDigits BinaryExponent FloatTypeSuffix_opt
-
-        val Number = "($Digits(\\.)?($Digits?)($Exp)?)|" +  // Digits ._opt Digits_opt ExponentPart_opt FloatTypeSuffix_opt
-                "(\\.($Digits)($Exp)?)|" +                  // . Digits ExponentPart_opt FloatTypeSuffix_opt
-                "(($HexString)[pP][+-]?$Digits)"            // HexSignificand BinaryExponent
-
-        val fpRegex = "[\\x00-\\x20]*[+-]?(NaN|Infinity|(($Number)[fFdD]?))[\\x00-\\x20]*"
-
-        Regex(fpRegex)
-    }
-}
-
 private inline fun <T> screenFloatValue(str: String, parse: (String) -> T): T? {
     return try {
-        if (ScreenFloatValueRegEx.value.matches(str))
+        if (isValidFloat(str))
             parse(str)
         else
             null
-    } catch (e: NumberFormatException) {  // overflow
+    } catch (_: NumberFormatException) {  // overflow
         null
     }
 }
+
+private const val LengthOfNaN = 2 // "NaN".length - 1
+private const val LengthOfInfinity = 7 // "Infinity".length - 1
+
+private fun isValidFloat(s: String): Boolean {
+    // A float can have one of two representations:
+    //
+    // 1. Standard:
+    //     - With an integer part only: 1234
+    //     - With integer and fractional parts: 1234.4678
+    //     - With a fractional part only: .4678
+    //
+    //     Optional sign prefix: + or -
+    //     Optional signed exponent: e or E, followed by optionally signed digits (+12, -12, 12)
+    //     Optional suffix: f, F, d, or D (for instance 12.34f or .34D)
+    //
+    // 2. Hexadecimal:
+    //     - With an integer part only: 0x12ab
+    //     - With integer and fractional parts: 0x12ab.CD78
+    //     - With a fractional part only: 0x.CD78
+    //
+    //     Mandatory signed exponent: p or P, followed by optionally signed digits (+12, -12, 12)
+    //
+    //     Optional sign prefix: + or -
+    //     Optional suffix: f, F, d, or D (for instance 12.34f or .34D)
+    //
+    // Two special cases:
+    //     "NaN" and "Infinity" strings, can have an optional sign prefix (+ or -)
+    //
+    // Implementation notes:
+    //     - The pattern "myChar.code or 0x20 == 'x'.code" is used to perform a case-insensitive
+    //       comparison of a character. Adding the 0x20 bit turns an upper case ASCII letter into
+    //       a lower case one.
+
+    var start = 0
+    var end = s.length - 1
+
+    // Skip leading spaces
+    while (start <= end && s[start].code <= 0x20) start++
```

Comment:
Introduction of something like `inline fun String.advanceWhile(index: Int, endInclusive: Int, predicate: (Char) -> Boolean): Int` will shorten the whole function and will improve its readability:

- `start = s.advanceWhile(start, end) { it.code <= 20 }`
- `start = s.advanceWhile(start, end) { it.isAsciiDigit() || it.isHexLetter() }`
- `start = s.advanceWhile(start, end) { it.isAsciiDigit() }`